### ISSUE TITLE
@JsonFormat doesn't work

### ISSUE THREAD

**Issue by belovaf** [yahoo/elide#634]
`@JsonFormat` on `Date` field doesn't work.
```
public class Entity {
    // Id and other fields declaration ...

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", timezone = "GMT")
    public Date someTimestamp;
}
```
Can you suggest any workaround?

**Comment by DennisMcWherter** [comment]
This is a known issue unfortunately-- our normalization step actually removes some of this metadata. It is something we do need to fix.

However, as a suggested work around, you can use your own custom serializer/deserializer to the object mapper. See this ticket for more information: https://github.com/yahoo/elide/issues/613

Let me know if that doesn't solve the issue!

**Comment by belovaf** [comment]
Thank you for your suggestion.
With configuration from #613 Elide returns formatted date, but accepts numeric timestamps.
Unfortunately, custom deserializer for Date is not called in this case.

**Comment by DennisMcWherter** [comment]
Sorry @belovaf, I was just referring to that as an example of using a custom serializer/deserializer. For your use case, I would expect you should be able to modify that code to write a new string out. Depending what kind of `Date` object you have, you should be able to do something like:

```java
...
new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").format(someTimestamp);
...
```

Is that not sufficient for some reason?

**Comment by belovaf** [comment]
@DennisMcWherter, thank you for response. Sure I'm able format `Date` to String. :)
The problem is that if I register serializer/deserializer, only serializer method executes.
```
mapper.getObjectMapper().registerModule(
  new SimpleModule("dateConverter", Version.unknownVersion())
    .addDeserializer(Date.class, new JsonDeserializer<Date>() { /* doesn't invoked on jsonpatch request */ })
    .addSerializer(Date.class, new JsonSerializer<Date>() { /* invoked on get request */ })
```

**Comment by DennisMcWherter** [comment]
I understand the problem now; I will have to investigate this a bit deeper when I have a chance. At the very least, serializers/deserializers should work. Hopefully can report back soon.

**Comment by sdhawley** [comment]
Having same problem, deserializer is not being called to convert json to Date.

Any ideas what is going on here?

**Comment by aklish** [comment]
@DennisMcWherter Did you ever get a chance to investigate this issue?

**Comment by aklish** [comment]
@sdhawley Do you have an example test case?

**Comment by sdhawley** [comment]
@aklish I used example code from issue #613 and tried posting data to a model that included a date field (java.util.Date) serialized as iso8601. It wasn't anything complex.

```java
@Include(rootLevel = true)
@Entity
public class Note extends ChildModel
{
    private static final long serialVersionUID = 1L;
    @Id 
    @GeneratedValue(strategy = GenerationType.IDENTITY) 
    public long id;
    @Lob
    @Column( length = 4096 )
    public String noteText = "";
        
    @Temporal( TemporalType.TIMESTAMP )
    public Date createTimeStamp;
    
    @Temporal( TemporalType.TIMESTAMP )
    public Date editTimeStamp;
    public Contact contact;
     
    @Override
    public long getId()
    {
        return id;
    }
}
```

**Comment by aklish** [comment]
Do you get an error like: ```{"errors":["InvalidValueException: Invalid value: 2018-07-04T03:45Z"]}```
?

**Comment by sdhawley** [comment]
Yes I was getting the InvalidValueException, so I tried to ad the custom deserializer but it never got called

**Comment by aklish** [comment]
I can reproduce and know what's going on.  I think there is a workaround that should work (need to try it first).   I hopefully will have some time tomorrow to do a writeup of how serde works and try out the workaround.  In a nutshell though, because Elide supports different languages:

- JSON-API entity bodies
- RSQL Filters
- GraphQL

deserialization ultimately doesn't rely on Jackson - it uses ```org.apache.commons.beanutils.ConvertUtils``` to coerce fields to types.

For JSON-API, Jackson is used to convert the entity body to intermediate representations (JsonApiDocument and Resource).   However, attributes are left as Objects - and the deserialization isn't smart enough (currently) to map that all the way to the underlying model classes (that happens later during type coercion).   While this could be fixed fairly easily, it doesn't solve the larger problem of the need to do type conversions in other places (where the incoming representation isn't JSON).  

The quick workaround likely involves registering a Converter with ConvertUtils.  That won't get
@JsonFormat working though.

I'll need to think more if there is an elegant solution that both supports model annotations like @JsonFormat _and_ allows deserialization from non-json formats without having to duplicate code.  Open to suggestions on that front.

**Comment by aklish** [comment]
I have a PR out that you can build against.  It has example code in Elide standalone for support for ISO8601 dates (and a change to Elide ConvertUtils that allows registration of converters).

To get this to work, it required an ObjectMapper serializer (from object->json) for JsonAPI and a BeanUtils Converter (from string->date) for filters & everything else.  ElideResourceConfig has the example of setting these up.

I haven't tested graphQL or added tests - so PR isn't done.  I tested POST, GET, and date comparison filters (>,<, etc).

Let me know if this workaround meets your needs.

**Comment by aklish** [comment]
PR #672 is released in Elide 4.2.5.  While `@JsonFormat` still won't work for dates, this fix should provide a more integrated solution for Date formatting that works in GraphQL & Filters (not just JSON serialization).  

You can enable different date formatting options via ElideSettings.  I'm going to close the issue and consider resolved unless I hear otherwise.
